core/src/apps/cardano/sign_message.py

```diff
@@ -0,0 +1,160 @@
+from micropython import const
+from typing import TYPE_CHECKING
+
+from trezor.wire import ProcessError
+from trezor.wire.context import call as ctx_call
+
+from apps.common import cbor
+
+from . import addresses, seed
+
+if TYPE_CHECKING:
+    from typing import Any
+
+    from trezor.messages import CardanoSignMessageFinished, CardanoSignMessageInit
+
+    from apps.common.cbor import CborSequence
+
+    Headers = dict[str | int, Any]
+
+_COSE_HEADER_ADDRESS_KEY = "address"
+_COSE_HEADER_ALGORITHM_KEY = const(1)
+_COSE_EDDSA_ALGORITHM_ID = const(-8)
+
+_MAX_CHUNK_SIZE = const(1024)
+
+
+def _get_chunks_count(size: int) -> int:
+    from math import ceil
+
+    return ceil(size / _MAX_CHUNK_SIZE)
+
+
+def _validate_chunk(
+    chunk_data: bytes,
+    chunk_index: int,
+    chunks_count: int,
+) -> None:
+    if len(chunk_data) > _MAX_CHUNK_SIZE:
+        raise ProcessError("Chunk too large")
+
+    if chunk_index < chunks_count - 1 and len(chunk_data) < _MAX_CHUNK_SIZE:
+        raise ProcessError("Intermediate chunk too small")
```

Comment:
Right, but in case of InlineDatums, the check that the total number of bytes is correct [is performed](https://github.com/trezor/trezor-firmware/blob/main/core/src/apps/cardano/helpers/hash_builder_collection.py#L63-L64) by `HashBuilderEmbeddedCBOR`. But I agree it's easy to miss - if we add a generic chunk manager, the correct length would be checked twice (in the chunk manager and in the hash builder) - but I think it's okay that way.

Reply:
Oh, you mean checking whether the chunks add up to the specified byte size. I'll definitely add that check to the generic chunk iterator that will be in the following commits